src/node_api_async.cc

```diff
@@ -1,69 +1,105 @@
 #include "node_api_async.h"
 #include "uv.h"
 
-typedef struct napi_work_impl__ {
-  uv_work_t* work;
-  void* data;
-  void (*execute)(void* data);
-  void (*complete)(void* data);
-  void (*destroy)(void* data);
-} napi_work_impl;
+namespace uvimpl {
 
-napi_work napi_create_async_work() {
-  napi_work_impl* worker =
-      reinterpret_cast<napi_work_impl*>(malloc(sizeof(napi_work_impl)));
-  uv_work_t* req = reinterpret_cast<uv_work_t*>(malloc(sizeof(uv_work_t)));
-  req->data = worker;
-  worker->work = req;
-  return reinterpret_cast<napi_work>(worker);
-}
+// Wrapper around uv_work_t which calls user-provided callbacks.
+class Work {
+ private:
+  explicit Work(napi_env env,
+                napi_async_callback execute = nullptr,
+                napi_async_callback complete = nullptr,
+                napi_async_callback destroy = nullptr,
+                void* data = nullptr)
+     : _env(env),
+     _data(data),
+     _execute(execute),
+     _complete(complete),
+     _destroy(destroy) {
+    _request = new uv_work_t;
+    _request->data = this;
+  }
 
-void napi_delete_async_work(napi_work w) {
-  napi_work_impl* worker = reinterpret_cast<napi_work_impl*>(w);
-  if (worker != NULL) {
-    if (worker->work != NULL) {
-      delete reinterpret_cast<uv_work_t*>(worker->work);
-    }
-    delete worker;
-    worker = NULL;
+  ~Work() { 
+	delete _request;
+  };
+
+ public:
+  static Work* New(napi_env env,
+                   napi_async_callback execute,
+                   napi_async_callback complete,
+                   napi_async_callback destroy,
+                   void* data) {
+    return new Work(env, execute, complete, destroy, data);
   }
-}
 
-void napi_async_set_data(napi_work w, void* data) {
-  napi_work_impl* worker = reinterpret_cast<napi_work_impl*>(w);
-  worker->data = data;
-}
+  static void Delete(Work* work) {
+    delete work;
+  }
 
-void napi_async_set_execute(napi_work w, void (*execute)(void* data)) {
-  napi_work_impl* worker = reinterpret_cast<napi_work_impl*>(w);
-  worker->execute = execute;
-}
+  static napi_status Queue(Work* work) {
+    // Consider: Choose a different loop based on the napi_env?
+    int result = uv_queue_work(uv_default_loop(),
```

Comment:
There isn't necessarily a relationship between the `napi_env` (`Isolate`) and libuv loops.

In order to support a non-default loop, N-API would have to either wrap most of [the libuv loop APIs](http://docs.libuv.org/en/v1.x/loop.html), or allow an optional `uv_loop_t` parameter to `napi_create_async_work()` (breaking the abstraction barrier). I don't really like either of those alternatives right now.

Another consideration is that the libuv docs [say](http://docs.libuv.org/en/v1.x/loop.html#c.uv_default_loop):

> default loop can (and should) be closed with uv_loop_close() so the resources associated with it are freed

and N-API currently doesn't offer any way to do that.